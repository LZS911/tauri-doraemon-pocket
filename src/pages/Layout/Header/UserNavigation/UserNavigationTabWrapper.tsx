import { UserNavigationTabWrapperProps } from '.';

const UserNavigationTabWrapper: React.FC<UserNavigationTabWrapperProps> = ({
  dataSource,
}) => {
  return (
    <div className="flex flex-col">
      {dataSource.map((v) => {
        return (
          !v.hidden && (
            <div
              key={v.text}
              onClick={v.onClick}
              className="flex cursor-pointer items-center rounded-sm px-4 py-2 transition-[background] hover:bg-slate-200 dark:text-white dark:hover:bg-slate-500"
            >
              {v.icon}
              <div className="ml-4">{v.text}</div>
            </div>
          )
        );
      })}
    </div>
  );
};

export default UserNavigationTabWrapper;
