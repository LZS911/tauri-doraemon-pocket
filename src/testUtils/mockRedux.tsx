import { configureStore } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';
import layout from '../store/layout';
import config from '../store/config';
import { Dictionary } from '../typing/common.type';
import { Provider } from 'react-redux';
import React, { ReactNode } from 'react';
const reducers = combineReducers({
  layout,
  config,
});
export const storeFactory: any = (initStore: Dictionary = {}) => {
  return configureStore({
    reducer: reducers,
    preloadedState: initStore,
  });
};

export const CustomProvider: React.FC<{
  initStore: Dictionary;
  children: ReactNode;
}> = (props) => {
  return (
    <Provider store={storeFactory(props.initStore)}>{props.children}</Provider>
  );
};
