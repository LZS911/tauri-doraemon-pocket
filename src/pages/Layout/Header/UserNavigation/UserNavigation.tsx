import {
  LogoutOutlined,
  SettingOutlined,
  UserOutlined,
} from '@ant-design/icons';
import { Popover, Tabs } from 'antd';
import { useTranslation } from 'react-i18next';
import {
  UserNavigationTabsKeyEnum,
  UserProfileTabDataSource,
  UserSettingsTabDataSource,
} from '.';
import ThemeBase from '../../../../components/ThemeBase';
import useUserConfig from '../../../../customHooks/useUserConfig';
import UserNavigationTabWrapper from './UserNavigationTabWrapper';

const UserNavigation: React.FC = () => {
  const { t } = useTranslation();
  const { username, userProfile, signOut } = useUserConfig();
  return (
    <Popover
      trigger="click"
      content={
        <div className="w-72">
          <header className="flex items-center justify-between p-6">
            <div className="flex">
              <img
                title="logo"
                className="h-8 w-8 rounded-[50%] bg-primary"
                src="https://joeschmoe.io/api/v1/random"
              />
              <div className="ml-3 font-bold">
                <div>{username}</div>
                <div className="text-xs opacity-50">{userProfile}</div>
              </div>
            </div>

            <ThemeBase.Icon
              data-testid="sign-out-icon"
              onClick={signOut}
              icon={<LogoutOutlined />}
              className="text-[1.25rem]"
            />
          </header>
          <div>
            <Tabs centered={true}>
              <Tabs.TabPane
                key={UserNavigationTabsKeyEnum.profile}
                tab={
                  <div className="flex items-center px-4">
                    <UserOutlined />
                    <span>
                      {t('layout.header.userNavigation.profile.title')}
                    </span>
                  </div>
                }
              >
                <UserNavigationTabWrapper
                  dataSource={UserProfileTabDataSource()}
                />
              </Tabs.TabPane>
              <Tabs.TabPane
                key={UserNavigationTabsKeyEnum.settings}
                tab={
                  <div className="flex items-center px-4">
                    <SettingOutlined />
                    <span>
                      {t('layout.header.userNavigation.settings.title')}
                    </span>
                  </div>
                }
              >
                <UserNavigationTabWrapper
                  dataSource={UserSettingsTabDataSource()}
                />
              </Tabs.TabPane>
            </Tabs>
          </div>
        </div>
      }
    >
      <ThemeBase.Icon
        data-testid="open-popover-icon"
        icon={
          <>
            <img
              title="logo"
              className="h-6 w-6 rounded-[50%] bg-primary"
              src="https://joeschmoe.io/api/v1/random"
            />
            <span className="ml-3 font-bold">{username}</span>
          </>
        }
        className="bg-slate-100 px-2 py-1 text-base dark:bg-black"
      />
    </Popover>
  );
};

export default UserNavigation;
