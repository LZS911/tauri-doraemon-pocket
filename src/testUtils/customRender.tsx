/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, renderHook } from '@testing-library/react';
import { Provider } from 'react-redux';
import {
  BrowserRouter,
  MemoryRouter,
  MemoryRouterProps,
  Router,
  RouterProps,
} from 'react-router-dom';
import { Dictionary } from '../typing/common.type';
import { storeFactory } from './mockRedux';
type RenderParams = Parameters<typeof render>;

export const renderWithRedux: (arg: any) => ReturnType<typeof render> = (
  ...[ui, option, initStore]: [...RenderParams, Dictionary?]
) => {
  return render(
    <Provider store={storeFactory(initStore)}>{ui}</Provider>,
    option
  );
};
export const renderWithRouter: any = (...[ui, option]: [...RenderParams]) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>, option);
};

export const renderHooksWithRedux = <TProps, TResult>(
  hooks: (props: TProps) => TResult,
  storeState: Dictionary = {}
) => {
  return renderHook(hooks, {
    wrapper: Provider,
    initialProps: {
      store: storeFactory(storeState),
    },
  } as any);
};

export const renderWithServerRouter: any = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(<Router {...props}>{ui}</Router>, option);
};

export const renderWithServerRouterAndRedux: any = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(
    <Router {...props}>
      {' '}
      <Provider store={storeFactory()}>{ui}</Provider>,
    </Router>,
    option
  );
};

export const renderWithMemoryRouter: any = (
  ...[ui, option, props]: [...RenderParams, MemoryRouterProps?]
) => {
  return render(<MemoryRouter {...props}>{ui}</MemoryRouter>, option);
};
