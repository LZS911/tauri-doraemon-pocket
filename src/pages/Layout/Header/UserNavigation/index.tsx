import {
  EditOutlined,
  FormOutlined,
  LockOutlined,
  MailOutlined,
  UserOutlined,
} from '@ant-design/icons';
import i18n from '../../../../locale';
import UserNavigation from './UserNavigation';
export interface ITabWrapperDataSource {
  icon: JSX.Element;
  onClick: () => void;
  text: string;
  hidden?: boolean;
}
export interface UserNavigationTabWrapperProps {
  dataSource: ITabWrapperDataSource[];
}
export enum UserNavigationTabsKeyEnum {
  profile,
  settings,
}
export const UserProfileTabDataSource: () => ITabWrapperDataSource[] = () => [
  {
    icon: <EditOutlined />,
    onClick: () => undefined,
    text: i18n.t('layout.header.userNavigation.profile.editProfile'),
  },
  {
    icon: <UserOutlined />,
    onClick: () => undefined,
    text: i18n.t('layout.header.userNavigation.profile.viewProfile'),
  },
  {
    icon: <MailOutlined />,
    onClick: () => undefined,
    text: i18n.t('layout.header.userNavigation.profile.bindEmail'),
  },
];
export const UserSettingsTabDataSource: () => ITabWrapperDataSource[] = () => [
  {
    icon: <UserOutlined />,
    onClick: () => undefined,
    text: i18n.t('layout.header.userNavigation.settings.accountSetting'),
  },
  {
    icon: <LockOutlined />,
    onClick: () => undefined,
    text: i18n.t('layout.header.userNavigation.settings.permissionSetting'),
  },
  {
    icon: <FormOutlined />,
    onClick: () => undefined,
    text: i18n.t('layout.header.userNavigation.profile.modifyPassword'),
  },
];

export default UserNavigation;
