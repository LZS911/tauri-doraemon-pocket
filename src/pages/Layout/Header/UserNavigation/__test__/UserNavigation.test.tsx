import { fireEvent, screen } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import { renderWithRedux } from '../../../../../testUtils/customRender';
import UserNavigation from '../index';

jest.mock('react-redux', () => {
  return {
    ...jest.requireActual('react-redux'),
    useDispatch: jest.fn(),
  };
});

describe('test UserNavigation', () => {
  const useDispatchMock = useDispatch as jest.Mock;
  const dispatchMock = jest.fn();

  beforeEach(() => {
    useDispatchMock.mockImplementation(() => dispatchMock);
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  test('should match snapshot', () => {
    const { container } = renderWithRedux(<UserNavigation />);
    fireEvent.click(screen.getByTestId('open-popover-icon'));
    expect(container).toMatchSnapshot();
  });
  test('should sign out when clicking sign icon', () => {
    renderWithRedux(<UserNavigation />);
    expect(dispatchMock).toBeCalledTimes(0);
    fireEvent.click(screen.getByTestId('open-popover-icon'));
    fireEvent.click(screen.getByTestId('sign-out-icon'));
    expect(dispatchMock).toBeCalledTimes(2);
    expect(dispatchMock).toBeCalledWith({
      type: 'userConfig/setLoginUserInfo',
      payload: {
        emailAddress: '',
        userProfile: '',
        username: '',
      },
    });

    expect(dispatchMock).toBeCalledWith({
      type: 'userConfig/setLoginState',
      payload: {
        isLogin: false,
        token: '',
      },
    });
  });
});
