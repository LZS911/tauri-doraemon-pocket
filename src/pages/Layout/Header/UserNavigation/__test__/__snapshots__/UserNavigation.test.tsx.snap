// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test UserNavigation should match snapshot 1`] = `
<div>
  <div
    class="hover:bg-slate-200 dark:hover:bg-slate-400 after:water-wave-hide active:after:water-wave-show bg-slate-100 px-2 py-1 text-base dark:bg-black ant-popover-open transparent relative flex cursor-pointer items-center justify-center rounded-md bg-transparent p-2 transition-colors dark:text-white"
    data-testid="open-popover-icon"
  >
    <img
      class="h-6 w-6 rounded-[50%] bg-primary"
      src="https://joeschmoe.io/api/v1/random"
    />
    <span
      class="ml-3 font-bold"
    />
  </div>
</div>
`;
